### ISSUE TITLE
JANA based event merger

### ISSUE THREAD

**Issue by simonge** [eic/EICrecon#1592]
In order to properly understand the physics performance of the ePIC detector we need to include hits from background sources in our reconstruction workflow. While the [HEPMC_Merger](https://github.com/eic/HEPMC_Merger) exists it is far from an ideal solution to require mixed events to be passed through the simulation. Instead, merging events between the simulation and digitization stage, allows the same simulated events to be used multiple times in different studies such as investigating how reconstruction is effected by luminosity, amongst many others.

Requirements of the merger

- Source events from multiple files
- Simultaneously access multiple events from the same file
- Distribute interaction events through time based on bunch number
- Distribute beam background events through time based on bunch number, bunch structure and position along beamline.

Main issues with any approach

- Keeping track of associations between MCParticles and hits 
- Collections with the same names being read from multiple files

This feels like it should be possible with the event folder, folding lots of physics events into a timeframe but from what I've seen that isn't quite right.

Potential approach (full of pitfalls)

- Load each source file tree independently
- In a new tree/frame create new renamed collections for each source which contains a reference to the simulation hit, time offset and event number
- Pass all of these new collections into the appropriate detector digitization method

**Comment by veprbl** [comment]
> This feels like it should be possible with the event folder, folding lots of physics events into a timeframe but from what I've seen that isn't quite right.

Can you expand on that?

> In a new tree/frame create new renamed collections for each source which contains a reference to the simulation hit, time offset and event number

There is no need to "rename" collections. It's just that we need to add functionality for PODIO Source to have modified tags, that are not simply original names of the collections in the frames.

**Comment by simonge** [comment]
Alternative more end user friendly approach is probably copy everything and do manual book keeping of the associations so it can all be kept in the same file

**Comment by veprbl** [comment]
DDG4 doesn't produce any associations. We only need to copy hits from different collections to a single one, plus modify their timestamps.

**Comment by simonge** [comment]
> DDG4 doesn't produce any associations. We only need to copy hits from different collections to a single one, plus modify their timestamps.

Sorry, I meant the relations rather than associations, OneToOneRelations from SimTrackerHits/CaloHitContribution to MCParticles and the OneToManyRelations between SimCalorimeterHit and CaloHitContribution.

**Comment by simonge** [comment]
> > This feels like it should be possible with the event folder, folding lots of physics events into a timeframe but from what I've seen that isn't quite right.
> 
> Can you expand on that?
> 

Despite the many presentations on the concept from @nathanwbrei I haven't been able to conceptually follow what happens to the data when moving between timeframe/event/sub-event. All of the examples start with a higher level and break it into smaller ones, which means they can be processed as a conventional event read and just have tags saying what event/subevent it belongs to within a larger collection. I would be more than happy for my understanding to be wrong and have it rewritten.

**Comment by veprbl** [comment]
(Replying to the last comment)

I might be wrong, but the division between timeframe, event, sub-event, while it is currently hardcoded (I believe, some constants are defined), is not implying an actual hierarchy of number of JEvents being passed. What I mean, is you could "fold" from timeframe to event and produce more events than timeframe, or "unfold" from timeframe to event and produce less events than timeframes. You can also do anything in between. And, I remember, @nathanwbrei mentioned that the level could, in principle be created by the users with arbitrary names (we could imaging "signal", "bg1", "bg2" instead), but for now we could (mis-)appropriate the existing defined levels for our needs.

**Comment by simonge** [comment]
Folding/unfolding actually changing the number of JEvents rather than tagging subsets of the event was my original impression. Keeping track of relations/associations between levels still sounds tricky.

My understanding is that a record in a podio collection can only ever belong to one collection, if you want to include it in another collection you need to either use a subset collection or copy it. Creating a subset collection which points to a different/several different JEvents wouldn't work (as far as I can see at least when writing out the objectID wouldn't work) while copying it will break any associations to it.

Here you'd want to have the record owned/shared by collections in JEvents on different levels. That way when saving to an output file with many frames/trees representing different levels both can contain the same records with their relations in tact.

This might of course be what's happening in which case, fantastic, I'm slowly catching up.

, possibly not too bad when in memory if you can directly point to the (e.g.) MCParticle but ideally this same MCParticle would belong to a collection in both the event and timeframe levels, the JEvent processor would know which level and collection the MCParticle  that writing out

**Comment by veprbl** [comment]
I don't think we can do without rewriting collection id's and indices. Objects will have to be copied in any case, as we need to modify time fields in most of them.

**Comment by simonge** [comment]
The alternative would be having an event data type with a time offset (along with some tag which says what the event source was) and adding a bunch of association collections between the hits and particles to that.

Will something like that not be needed to trace back through the event levels anyway?

**Comment by simonge** [comment]
Current thoughts on a process which as far as I know would work but needs access to some additional JANA features:

For each event source j

- Create a timeline to inject N_i events into for each sample
- Use the JANA EventGroup to collect N_i events.
- For each i event:

1. Create a new edm4eic::EventSource with a time offset and identifier, this should either be a one to many or one to one relation with the MCParticles.
2. Copy the MCParticles, offsetting the times by the timeline time and creating an association with the original MCParticle.
3. Loop over the associations updating the parent/daughter fields of the new MCParticles.
4. Copy the SimTrackerHits, offsetting the time and updating the relation based on the association.
5. Copy the CaloHitContribution, offsetting the time and updating the relation based on the association. Create an association to the original CaloHitContribution.
6. Copy the SimCalorimeterHits, updating their relation based on the contribution associations.

Bringing the event sources together is a bit simpler and could probably be handled in a number of different ways depending on whether it is attached to the digitization/reconstruction/stand alone. Probably easiest would be a stand alone merger (or at least separate JANA Plugin if that's how they're meant to work) where each source has 4 collections with a source tag then original named collections are just a subset collection with those merged.

Merging of the metadata/checking for conflicts is another kettle of fish.

**Comment by simonge** [comment]
@nathanwbrei would you be able to comment on this and direct me on how to get started.

**Comment by nathanwbrei** [comment]
@simonge Let's schedule a call!

**Comment by simonge** [comment]
I have made a stand alone JANA plugin, (using the most recent JANA version not in eic_shell), which can simply unfold events from one level to build up a collection in another. This could be used to create a file with timeslices for a single event source which get merged with other sources in a separate plugin/eicrecon.

Based on digging into JANA over the last few days I believe it currently isn't possible to combine the event sources as we would require in a single instance of JANA. From my understanding:

- Sourcing events at the same `JEventLevel` puts them in the same event, so the unfolder wouldn't be able to mix various quantities of each.
- Sourcing the events into different `JEventLevels` and providing an unfolder into the level of the output isn't supported. I naïvely tried to just rearrange the `JTopologyBuilder` to ensure it loops through all event levels and makes the connections properly (see https://github.com/simonge/JANA2/blob/simong-source-playground/src/libraries/JANA/Topology/JTopologyBuilder.cc) but the scope of changes required appears to be more significant than this alone, (or understand if this approach actually would work).

@nathanwbrei Is this what your `nbrei_multilevel_sources` branch is trying to address?

**Comment by simonge** [comment]
There is probably quite a lot which could be loaded from the `podio_metadata` which would simplify the code and probably allow for adding the names of the association branches directly rather than hard coding.

If it becomes possible to handle this directly in JANA/EICrecon (or future EICdigi) when the overheads of building the podio frame from the data will be needed anyway then we could move back to using the relations directly.

I hadn't realised quite how slow the simulation is at the moment with about 1Hz per background bremsstrahlung electron and naturally slower for other samples. So as soon as a simulated event is used more than once it will be worth merging after rather than before.
